drape/glyph_manager.cpp

```diff
@@ -659,46 +626,68 @@ hb_language_t OrganicMapsLanguageToHarfbuzzLanguage(int8_t lang)
 
+// This method is NOT multithreading-safe.
 text::TextMetrics GlyphManager::ShapeText(std::string_view utf8, int fontPixelHeight, int8_t lang)
 {
+#ifdef DEBUG
+  static int const fontSize = fontPixelHeight;
+  ASSERT_EQUAL(fontSize, fontPixelHeight,
+      ("Cache relies on the same font height/metrics for each glyph", fontSize, fontPixelHeight));
+#endif
+
+  // A simple cache greatly speeds up text metrics calculation. It has 80+% hit ratio in most scenarios.
+  if (auto const found = m_impl->m_textMetricsCache.find(utf8); found != m_impl->m_textMetricsCache.end())
+    return found->second;
+
   const auto [text, segments] = harfbuzz_shaping::GetTextSegments(utf8);
 
   // TODO(AB): Optimize language conversion.
   hb_language_t const hbLanguage = OrganicMapsLanguageToHarfbuzzLanguage(lang);
 
   text::TextMetrics allGlyphs;
 
-  // TODO(AB): Cache runs.
-  // TODO(AB): Cache buffer.
-  hb_buffer_t * buf = hb_buffer_create();
   for (auto const & substring : segments)
   {
-    hb_buffer_clear_contents(buf);
+    hb_buffer_clear_contents(m_impl->m_harfbuzzBuffer);
 
     // TODO(AB): Some substrings use different fonts.
-    hb_buffer_add_utf16(buf, reinterpret_cast<const uint16_t *>(text.data()), static_cast<int>(text.size()),
-                        substring.m_start, substring.m_length);
-    hb_buffer_set_direction(buf, substring.m_direction);
-    hb_buffer_set_script(buf, substring.m_script);
-    hb_buffer_set_language(buf, hbLanguage);
-
-    // TODO(AB): Check not only the first character to determine font for the run, but all chars.
-    auto firstCharacterIter{text.begin() + substring.m_start};
-    auto const firstCharacterUnicode = utf8::unchecked::next16(firstCharacterIter);
-
-    int const fontIndex = GetFontIndex(firstCharacterUnicode);
-    if (fontIndex < 0)
+    hb_buffer_add_utf16(m_impl->m_harfbuzzBuffer, reinterpret_cast<const uint16_t *>(text.data()),
+        static_cast<int>(text.size()), substring.m_start, substring.m_length);
+    hb_buffer_set_direction(m_impl->m_harfbuzzBuffer, substring.m_direction);
+    hb_buffer_set_script(m_impl->m_harfbuzzBuffer, substring.m_script);
+    hb_buffer_set_language(m_impl->m_harfbuzzBuffer, hbLanguage);
+
+    auto u32CharacterIter{text.begin() + substring.m_start};
+    auto const end{text.begin() + substring.m_start + substring.m_length};
+    do
     {
-      // TODO(AB): Add missing glyph character's metrics
-      LOG(LWARNING, ("Skip run because no font was found for character", NumToHex(firstCharacterUnicode)));
-      continue;
-    }
+      auto const u32Character = utf8::unchecked::next16(u32CharacterIter);
+      // TODO(AB): Mapping characters to fonts can be optimized.
+      int const fontIndex = GetFontIndex(u32Character);
+      if (fontIndex < 0)
+        LOG(LWARNING, ("No font was found for character", NumToHex(u32Character)));
+      else
+      {
+        // TODO(AB): Mapping font only by the first character in a string may fail in theory in some cases.
```

Comment:
Ok, why TODO then? :)

Reply:
Because implementing the font lookup fallback is complex and likely unnecessary here.

Segmenting text into scripts already implies that all or most characters from this script should be present in the font. This failure case is unlikely, and in the worst case, some unrendered glyphs will be visible. So we can come back and debug it.